### PR TITLE
ARROW-2494: [C++] Return status codes from PlasmaClient::Seal instead of crashing

### PR DESCRIPTION
**Pull request by kszucs**

### CODE REVIEW

File: cpp/src/plasma/client.cc [hunk 1]
```diff
@@ -604,10 +604,16 @@ Status PlasmaClient::Seal(const ObjectID& object_id) {
...
   // request to Plasma.
   auto object_entry = objects_in_use_.find(object_id);
-  ARROW_CHECK(object_entry != objects_in_use_.end())
-      << "Plasma client called seal an object without a reference to it";
-  ARROW_CHECK(!object_entry->second->is_sealed)
-      << "Plasma client called seal an already sealed object";
+
+  if (object_entry == objects_in_use_.end()) {
+    return Status::PlasmaObjectNonexistent(
+        "Plasma client called seal an object without a reference to it");
```
Comment (pitrou) [inline cpp/src/plasma/client.cc]: I'm not a native English speaker, but while we're at it, perhaps we could fix the grammar in this error message and the one below?
Something like "Seal() called on an object without a reference to it".

File: cpp/src/plasma/test/client_tests.cc [hunk 2]
```diff
@@ -90,6 +90,21 @@ class TestPlasmaStore : public ::testing::Test {
   PlasmaClient client2_;
 };
 
+TEST_F(TestPlasmaStore, SealErrorsTest) {
+  ObjectID object_id = ObjectID::from_random();
+
+  Status result = client_.Seal(object_id);
+  ASSERT_EQ(result.IsPlasmaObjectNonexistent(), true);
```
Comment (pitrou) [inline cpp/src/plasma/test/client_tests.cc]: You can use ASSERT_TRUE().

File: cpp/src/plasma/test/client_tests.cc [hunk 3]
```diff
@@ -90,6 +90,21 @@ class TestPlasmaStore : public ::testing::Test {
...
+
+  // Create object.
+  std::vector<uint8_t> data(100, 0);
+  CreateObject(client_, object_id, {42}, data);
+
+  // Trying to seal it again.
+  result = client_.Seal(object_id);
+  ASSERT_EQ(result.IsPlasmaObjectAlreadySealed(), true);
```
Comment (pitrou) [inline cpp/src/plasma/test/client_tests.cc]: Same here.

**Review by pitrou — COMMENTED** [review]
No real issues, but perhaps you want to address the small nits below.

File: cpp/src/plasma/client.cc (hunk 1 above)
Comment (kszucs) [reply cpp/src/plasma/client.cc]: Me neither :) Shouldn't we incorporate the object ID in the message?

**Comment by kszucs** [comment]
@pitrou I've created a subtask in JIRA. Generally We should replace all occurrences of `ARROW_CHECK`, right?

File: cpp/src/plasma/client.cc (hunk 1 above)
Comment (pitrou) [reply cpp/src/plasma/client.cc]: If there's a function to pretty-print object IDs when why not :-)

**Comment by pitrou** [comment]
When `ARROW_CHECK` checks for an internal invariant, we can keep it. But in other cases (such as non-existent object ID or IO error), it would be better to replace it with an error return, yes.

File: cpp/src/plasma/client.cc (hunk 1 above)
Comment (kszucs) [reply cpp/src/plasma/client.cc]: @pitrou I guess pretty printing of plasma object should follow the API of [arrow::pretty_print](https://github.com/apache/arrow/blob/master/cpp/src/arrow/pretty_print.h), if so should I create a JIRA ticket instead?

Comment (pitrou) [reply cpp/src/plasma/client.cc]: I'm not sure, but that sounds reasonable, and can be discussed on a JIRA issue indeed.

**Review by pitrou — APPROVED** [review]
LGTM.

**Pull request closed without merging**